bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/models/v3/product/ProductImpl.java

```diff
@@ -153,4 +183,111 @@ protected Variant mapVariant(ConfigurableVariant variant) {
+                    jsonArray.put(new JSONObject(variantMapWithNoSpecialPrice));  // Add to the array
+                } else {
+                    // Case when there's a special price
+                    variantMap.put("availability", variant.getInStock() ? "InStock" : "OutOfStock");
+
+                    // Price specification
+                    JSONObject priceSpecification = new JSONObject();
+                    priceSpecification.put("@type", "UnitPriceSpecification");
+                    priceSpecification.put("priceType", "https://schema.org/ListPrice");
+                    if (priceRange != null) {
+                        priceSpecification.put("price", priceRange.getRegularPrice());
+                        priceSpecification.put("priceCurrency", priceRange.getCurrency());
+                    }
+                    variantMap.put("priceSpecification", priceSpecification);
+
+                    variantMap.put("price", variantImpl.getSpecialPrice());  // Access special price directly
+                    variantMap.put("SpecialPricedates", variantImpl.getSpecialToDate());  // Access special date directly
+
+                    // Add the variant with special price to the array
+                    jsonArray.put(new JSONObject(variantMap));
+                }
+            }
+        }
+
+        return jsonArray;
+    }
+
+    @Override
+    public String generateProductJsonLDString() {
+        try {
+            // If JSON generation is disabled, return null early
+            if (!enableJson) {
+                return null;
+            }
+
+            // Initialize ObjectMapper to create JSON nodes
+            ObjectMapper mapper = new ObjectMapper();
+            ObjectNode productJson = mapper.createObjectNode();
+
+            // Set basic product attributes
+            productJson.put("@context", "http://schema.org");
+            productJson.put("@type", "Product");
+            productJson.put("sku", getSku());
+            productJson.put("name", getName());
+
+            // Assuming the first asset is the image
+            if (!getAssets().isEmpty()) {
+                productJson.put("image", getAssets().get(0).getPath());
+            } else {
+                productJson.put("image", "");  // Default to empty if no assets are available
+            }
+
+            productJson.put("description", getDescription());
+            productJson.put("@id", getId());
+
+            // Create the "offers" array for variants
+            ArrayNode offersArray = mapper.createArrayNode();
+            JSONArray offers = fetchVariantsAsJsonArray();
+            for (int i = 0; i < offers.length(); i++) {
+                offersArray.add(mapper.readTree(offers.get(i).toString()));
+            }
+            productJson.set("offers", offersArray);
+
+            // Return the JSON string representation
+            return mapper.writeValueAsString(productJson);
```

Comment:
Please save the json-ld in s field after generating it and when the method is called second time return the already generated json-ld from the field, so that you can avoid the duplicate generation of json-ld when you call the method two times in the same script (first for the  null check and then for adding the json-ld in the script tag).